### PR TITLE
Refactor MatrixWrite and use in Python

### PR DESCRIPTION
**Pull request by jigold**

### CODE REVIEW

File: hail/src/test/scala/is/hail/expr/ir/MatrixIRSuite.scala [hunk 1]
```diff
@@ -273,4 +273,35 @@ class MatrixIRSuite extends SparkSuite {
...
     }
   }
+
+  @Test def testMatrixWrite() {
+    // Native Writer
+    val range = MatrixTable.range(hc, 11, 3, Some(10))
+    var path = tmpDir.createLocalTempFile(extension = "mt")
+    Interpret(MatrixWrite(range.ast, MatrixNativeWriter(path)))
+    val read = MatrixTable.read(hc, path)
+    assert(read.same(range))
```
Comment (danking) [inline hail/src/test/scala/is/hail/expr/ir/MatrixIRSuite.scala]: Let's use four separate test methods so we have more fine-grained test success/failure information. you can make the `MatrixTable.range...` a class-level `lazy val`.

File: hail/src/test/scala/is/hail/expr/ir/MatrixIRSuite.scala [hunk 2]
```diff
@@ -273,4 +273,35 @@ class MatrixIRSuite extends SparkSuite {
...
+      FastIndexedSeq("varid" -> GetField(Ref("va", plink.rvRowType), "rsid"))))
+    path = tmpDir.createLocalTempFile()
+    Interpret(MatrixWrite(plinkIR, MatrixPLINKWriter(path)))
+
+    // Export GEN
+    val gen = hc.importGen("src/test/resources/example.gen", "src/test/resources/example.sample",
+      contigRecoding = Some(Map("01" -> "1")))
+    val genIR = MatrixMapCols(gen.ast, SelectFields(InsertFields(Ref("sa", gen.colType),
+      FastIndexedSeq("id1" -> Str(""), "id2" -> Str(""), "missing" -> F64(0.0))), FastIndexedSeq("id1", "id2", "missing")),
+      Some(FastIndexedSeq("id1"))) // FIXME: Make importGen add sample annotations from the sample file
```
Comment (danking) [inline hail/src/test/scala/is/hail/expr/ir/MatrixIRSuite.scala]: I think this is not a great place for this FIXME. Let's make it an issue and assign it to someone.

File: hail/src/test/scala/is/hail/expr/ir/MatrixIRSuite.scala [hunk 3]
```diff
@@ -273,4 +273,35 @@ class MatrixIRSuite extends SparkSuite {
...
+
+    // Export VCF
+    val vcf = hc.importVCF("src/test/resources/sample.vcf")
+    path = tmpDir.createLocalTempFile(extension = "vcf")
+    Interpret(MatrixWrite(vcf.ast, MatrixVCFWriter(path)))
+
+    // Export PLINK
+    val plinkPath = "src/test/resources/skip_invalid_loci"
+    val plink = hc.importPlink(plinkPath + ".bed", plinkPath + ".bim", plinkPath + ".fam", skipInvalidLoci = true)
+    val plinkIR = MatrixMapRows(plink.ast, InsertFields(Ref("va", plink.rvRowType), // FIXME: Make importPlink output varid as field name instead of rsid
```
Comment (danking) [inline hail/src/test/scala/is/hail/expr/ir/MatrixIRSuite.scala]: I think this FIXME is gonna get lost, let's create an issue and assign it to someone.

File: hail/src/main/scala/is/hail/expr/ir/MatrixValue.scala [hunk 4]
```diff
@@ -35,6 +35,39 @@ case class MatrixValue(
...
+          fatal(s"in $method: column key must be type 'str', found: ${t.mkString("[",",","]")}")
+      }
+    }
+
+    def referenceGenome: ReferenceGenome = {
+      val firstKeyField = typ.rowKeyStruct.types(0)
+      firstKeyField match {
+        case TLocus(rg: ReferenceGenome, _) => rg
+      }
+    }
```
Comment (danking) [inline hail/src/main/scala/is/hail/expr/ir/MatrixValue.scala]: These are all operations on `MatrixType`. If we move this code there, then `MatrixValue` and `MatrixTable` can share the implementation.

File: hail/src/test/scala/is/hail/expr/ir/MatrixIRSuite.scala (hunk 3 above)
Comment (tpoterba) [reply hail/src/test/scala/is/hail/expr/ir/MatrixIRSuite.scala]: this is a breaking change, we can't do it until 0.3.

File: hail/src/main/scala/is/hail/io/vcf/ExportVCF.scala [hunk 5]
```diff
@@ -199,13 +200,20 @@ object ExportVCF {
...
-    val tg = vsm.entryType match {
+    ExportVCF(mt.value, path, append, exportType, metadata)
+  }
+
+  def apply(mv: MatrixValue, path: String, append: Option[String],
+    exportType: Int, metadata: Option[VCFMetadata]) {
+
+    val typ = mv.typ
+
+    mv.requireColKeyString("export_vcf")
```
Comment (tpoterba) [inline hail/src/main/scala/is/hail/io/vcf/ExportVCF.scala]: these are checked in Python. We can just assert here

File: hail/src/main/scala/is/hail/expr/ir/MatrixValue.scala [hunk 6]
```diff
@@ -35,6 +35,39 @@ case class MatrixValue(
...
     }
 
+    def stringSampleIds: IndexedSeq[String] = {
+      val colKeyTypes = typ.colKeyStruct.types
+      assert(colKeyTypes.length == 1 && colKeyTypes(0).isInstanceOf[TString], colKeyTypes.toSeq)
+      val querier = typ.colType.query(typ.colKey(0))
+      colValues.value.map(querier(_).asInstanceOf[String])
+    }
+
+    def requireRowKeyVariant(method: String) {
```
Comment (tpoterba) [inline hail/src/main/scala/is/hail/expr/ir/MatrixValue.scala]: These checks are already done in Python - why do we need them in the JVM?

The python ones are way more readable (no java stack trace)

File: hail/src/main/scala/is/hail/io/vcf/ExportVCF.scala (hunk 5 above)
Comment (jigold) [reply hail/src/main/scala/is/hail/io/vcf/ExportVCF.scala]: These functions are doing the assertion. I think leaving this is fine.

Comment (jigold) [reply hail/src/main/scala/is/hail/io/vcf/ExportVCF.scala]: I modified them to just do an assert.

**Pull request merged**